File: apps/mobile/lib/custom_code/utils/worktree_selection.dart
// New-session worktree selection state + the spawn-arg decision logic.
// See `plans/todos/vicoa-app-worktree.md` §5.3.

/// How the new session relates to a git worktree:
///   - [none]        spawn in the chosen directory (today's behavior)
///   - [newWorktree] daemon forks a fresh branch+checkout off HEAD
///   - [existing]    spawn into an already-existing worktree
enum WorktreeMode { none, newWorktree, existing }

/// The resolved spawn-session inputs for a worktree selection: the directory to
/// run in, and the optional `worktree` RPC param.
typedef WorktreeSpawn = ({String directory, Map<String, dynamic>? worktree});

/// Map a [WorktreeMode] selection onto spawn-session args.
///
/// `newWorktree` keeps [baseDirectory] (the daemon forks off its HEAD) and asks
/// for creation; `existing` spawns directly in the selected worktree path with
/// no `worktree` param; `none` is today's plain spawn. An `existing` selection
/// with no path falls back to the base directory so a stale selection can never
/// produce an empty directory.
WorktreeSpawn resolveWorktreeSpawn({
  required WorktreeMode mode,
  required String baseDirectory,
  String? selectedWorktreePath,
}) {
  switch (mode) {
    case WorktreeMode.none:
      return (directory: baseDirectory, worktree: null);
    case WorktreeMode.newWorktree:
      return (directory: baseDirectory, worktree: {'new': true});
    case WorktreeMode.existing:
      return (
        directory: selectedWorktreePath ?? baseDirectory,
        worktree: null,
      );
  }
}

/// Session statuses that mean a session is still live. Mirrors the active set
/// used by the home session list; anything else (e.g. COMPLETED) is terminal.
const Set<String> kActiveSessionStatuses = {
  'ACTIVE',
  'AWAITING_INPUT',
  'REVIEWED',
};

/// Whether [path] looks like a vicoa-managed worktree (under
/// `~/vicoa/workspaces/`). A heuristic used only to decide whether to OFFER
/// cleanup — the daemon is the real authority and re-validates on remove.
bool isManagedWorktreePath(String path) =>
    path.contains('/vicoa/workspaces/');

/// Collapse an absolute home-dir path to `~` form (e.g. `/Users/dev/x` →
/// `~/x`). The daemon expands paths when it runs git, so `git worktree list`
/// returns absolute paths, while the CLI records a session's `project`
/// home-collapsed (`~/...`). Normalizing to the same `~` form lets the two line
/// up for comparison AND gives a tidier path to display. Returns [path]
/// unchanged (trailing slash trimmed) when [homeDir] is null/empty or doesn't
/// prefix it (already-relative or non-home paths).
String relativizeHome(String path, String? homeDir) {
  var p = path.trim();
  while (p.length > 1 && p.endsWith('/')) {
    p = p.substring(0, p.length - 1);
  }
  if (homeDir == null || homeDir.isEmpty) return p;
  var h = homeDir.trim();
  while (h.length > 1 && h.endsWith('/')) {
    h = h.substring(0, h.length - 1);
  }
  if (p == h) return '~';
  if (p.startsWith('$h/')) return '~${p.substring(h.length)}';
  return p;
}

/// Whether any session in [sessions] is still live AND running in
/// [worktreePath]. The daemon is a dumb executor with no session knowledge, so
/// the app must enforce "don't remove a worktree out from under a live session"
/// before calling `git-worktree-remove` (§5.5). [sessions] are raw agent
/// instance maps (`{'project': ..., 'status': ...}`); malformed entries are
/// ignored. [homeDir] (the machine's home) is required to match the absolute
/// git worktree path against the `~`-form session `project`; without it the
/// comparison falls back to exact string match.
bool worktreeHasActiveSession(String worktreePath, List<dynamic> sessions,
    {String? homeDir}) {
  final target = relativizeHome(worktreePath, homeDir);
  for (final s in sessions) {
    if (s is! Map) continue;
    final project = s['project'];
    final status = s['status'];
    if (project is! String || status is! String) continue;
    if (relativizeHome(project, homeDir) == target &&
        kActiveSessionStatuses.contains(status)) {
      return true;
    }
  }
  return false;
}

File: apps/web/lib/worktree-selection.ts
/**
 * New-session worktree selection logic for the web dashboard. Mirrors the
 * mobile `worktree_selection.dart`. Pure functions — unit-tested in
 * `worktree-selection.test.ts`.
 *
 * See the daemon side in vicoa-backend `rpc/worktree_*.py` and the plan
 * `plans/todos/vicoa-app-worktree.md`.
 */

import { joinSubpath } from '@/lib/project-paths';

/** How a new session relates to a git worktree. */
export type WorktreeMode = 'none' | 'new' | 'existing';

type MetadataShape = {
  metadata?: Record<string, unknown> | null;
  machine_metadata?: Record<string, unknown> | null;
};

function readCapabilities(machine: MetadataShape | null | undefined): unknown {
  if (!machine) return undefined;
  const meta = machine.metadata ?? machine.machine_metadata;
  if (!meta || typeof meta !== 'object') return undefined;
  return (meta as Record<string, unknown>).capabilities;
}

/**
 * Whether the machine's daemon advertises git-worktree support via
 * `metadata.capabilities`. Unlike agent availability, a MISSING capability
 * reads as UNsupported (an old daemon silently ignores the spawn-session
 * `worktree` param), so the UI hides the option unless it's explicitly there.
 */
export function machineSupportsWorktree(
  machine: MetadataShape | null | undefined,
): boolean {
  const caps = readCapabilities(machine);
  return Array.isArray(caps) && caps.some((c) => String(c) === 'worktree');
}

/**
 * Whether the daemon honours `worktree.name` on spawn and serves the
 * `git-worktree-check-name` RPC. Gated separately from `worktree`: an old
 * daemon drops the unknown `name` silently and spawns a random slug, so
 * offering the field against one would hand the user a different name than
 * they typed.
 */
export function machineSupportsWorktreeName(
  machine: MetadataShape | null | undefined,
): boolean {
  const caps = readCapabilities(machine);
  return Array.isArray(caps) && caps.some((c) => String(c) === 'worktree-name');
}

/**
 * Instant client-side verdict on a typed worktree name — the common
 * `git check-ref-format --branch` rules, so obvious mistakes are flagged as
 * the user types without a daemon round trip. Returns a short reason, or
 * null when the name passes. The daemon (`git-worktree-check-name`) stays the
 * authority: it runs git's real check and knows what's taken.
 */
export function validateWorktreeName(name: string): string | null {
  const n = name.trim();
  if (!n) return null; // blank = random name, always fine
  if (/\s/.test(n)) return 'No spaces';
  if (/[~^:?*[\\\u0000-\u001f\u007f]/.test(n)) return 'No ~ ^ : ? * [ or \\';
  if (n.startsWith('-')) return "Can't start with -";
  if (n.includes('..') || n.includes('@{') || n === '@') return 'No .. or @{';
  if (n.endsWith('.')) return "Can't end with .";
  // `feat/login` is a fine branch (nested middle dir); each segment must
  // still be non-empty, not dot-led, and not end in `.lock`.
  const parts = n.split('/');
  if (parts.some((part) => !part || part.startsWith('.') || part.endsWith('.lock'))) {
    return 'Bad path segment (empty, leading . or .lock)';
  }
  return null;
}

export interface WorktreeSpawn {
  directory: string;
  /** The spawn-session `worktree` param, or undefined for a plain spawn. */
  worktree: { new: true; name?: string } | undefined;
}

/**
 * Map a worktree selection onto spawn-session args. `baseDirectory` is the
 * folder the user picked — the repo root or a subfolder of it (a monorepo
 * session at `repo/apps/web`); `subpath` is that folder's part below the
 * root (`''` at the root). `new` sends the folder as-is (the daemon forks the
 * whole repo off its HEAD, on `newWorktreeName` when the user picked one, else
 * a random slug, and starts the agent at the same subfolder inside the new
 * checkout); `existing` starts at the selected worktree — at the same
 * subfolder inside it, so the checkout and the folder stay orthogonal;
 * `none` is a plain spawn. A stale `existing` selection with no path falls
 * back to the base directory.
 */
export function resolveWorktreeSpawn({
  mode,
  baseDirectory,
  subpath = '',
  selectedWorktreePath,
  newWorktreeName,
}: {
  mode: WorktreeMode;
  baseDirectory: string;
  subpath?: string;
  selectedWorktreePath?: string | null;
  newWorktreeName?: string | null;
}): WorktreeSpawn {
  switch (mode) {
    case 'new': {
      const name = newWorktreeName?.trim();
      return {
        directory: baseDirectory,
        worktree: name ? { new: true, name } : { new: true },
      };
    }
    case 'existing':
      return {
        directory: selectedWorktreePath
          ? joinSubpath(selectedWorktreePath, subpath)
          : baseDirectory,
        worktree: undefined,
      };
    case 'none':
    default:
      return { directory: baseDirectory, worktree: undefined };
  }
}

/**
 * Whether `path` looks like a vicoa-managed worktree (under
 * `~/vicoa/workspaces/`). A heuristic for "should we offer removal" — the
 * daemon re-validates on remove.
 */
export function isManagedWorktreePath(path: string): boolean {
  return path.includes('/vicoa/workspaces/');
}
